### PR TITLE
WIP Adding wallet utils and authentication

### PR DESCRIPTION
**Pull request by nitsujlangston**
# Goals:

- Provide client library and CLI utils for wallet operations (able to handle millions of imported keys in wallet client storage)
- Stop using mongo id as wallet id
- Authenticate wallet requests using payload signatures, signed by request auth key (derived from xPubKey on set path)

### CODE REVIEW

File: packages/bitcore-node/lib/routes/wallet.js [hunk 1]
```diff
@@ -2,27 +2,62 @@ const util = require('util');
...
 const _ = require('underscore');
+const secp256k1 = require('secp256k1');
 
 const logger = require('../logger');
 const ChainStateProvider = require('../providers/chain-state');
+const bitcoreLib = require('bitcore-lib');
+
+const verifyRequestSignature = (params) => {
+  const { message, pubKey, signature} = params;
+  const pub = (new bitcoreLib.HDPublicKey(pubKey)).deriveChild('m/2').publicKey.toString('hex');
```
Comment (nitsujlangston) [inline packages/bitcore-node/lib/routes/wallet.js]: Need to remove these toStrings - they were from debugging

**Comment by micahriggan** [comment]
Initial glance over seems good. I like the authenticate middleware. 

Is the idea that the new bitcore-client thing will replace bitcore-wallet? 

As for the converting from (arg, arg, arg) to (params), I think that's cool, but in the future we're going to have to read the code to see what's being pulled off of params.

Perhaps we should write the new stuff in TypeScript. I think it'd be beneficial.

File: packages/bitcore-node/lib/routes/wallet.js (hunk 1 above)
Comment (nitsujlangston) [reply packages/bitcore-node/lib/routes/wallet.js]: fixed

File: packages/bitcore-node/server.js [hunk 2]
```diff
@@ -25,7 +25,7 @@ async.series([
     await Promise.all(p2pServices.map(p2pService => p2pService.start()));
   }
 ], function () {
-  if (cluster.isWorker) {
+  if (!cluster.isWorker) {
```
Comment (micahriggan) [inline packages/bitcore-node/server.js]: Debug line

File: packages/bitcore-node/lib/routes/wallet.js [hunk 3]
```diff
@@ -2,27 +2,62 @@ const util = require('util');
...
   }
 });
 
-router.get('/:walletId', async function(req, res) {
+router.get('/:walletId', authenticate, async function(req, res) {
   try {
-    let {chain, network, walletId} = req.params;
-    let wallet = await ChainStateProvider.getWallet(chain, network, walletId);
+    let {chain, network, pubKey} = req.params;
+    let wallet = await ChainStateProvider.getWallet({ chain, network, pubKey });
```
Comment (micahriggan) [inline packages/bitcore-node/lib/routes/wallet.js]: Wouldn't you have req.wallet here?

File: packages/bitcore-client/bin/wallet-import [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  .option('--path <path>', 'REQUIRED - Where wallet is stored')
+  .parse(process.argv);
+
+const main = async () => {
+  const {path} = program;
+  let wallet;
+  try {
+    wallet = await Wallet.loadWallet({ path });
+    const keyInput = [
+      { address: 'mhse1Lz4otzATQWXGo7bGp2apbQGubvnjE', privKey: 'cNZPB2AgNpNjcLQQ9zjuCGgPBGP2ZNyNav6JLVHPUvHBByKruHT9' },
```
Comment (micahriggan) [inline packages/bitcore-client/bin/wallet-import]: Debug lines

File: packages/bitcore-node/lib/routes/wallet.js [hunk 5]
```diff
@@ -34,33 +69,23 @@ router.get('/:walletId', async function(req, res) {
 
 router.get('/:walletId/addresses', async function(req, res) {
```
Comment (micahriggan) [inline packages/bitcore-node/lib/routes/wallet.js]: Needs to have pubKey here

File: packages/bitcore-node/lib/routes/wallet.js [hunk 6]
```diff
@@ -150,10 +175,10 @@ ListTransactionsStream.prototype._transform = function(transaction, enc, done) {
   done();
 };
 
-router.get('/:walletId/transactions', async (req, res) => {
-  let {walletId, chain, network} = req.params; 
+router.get('/:pubKey/transactions', async (req, res) => {
```
Comment (micahriggan) [inline packages/bitcore-node/lib/routes/wallet.js]: Why doesn't this have authenticate, but the utxos does?

Comment (nitsujlangston) [reply packages/bitcore-node/lib/routes/wallet.js]: it should, all wallet methods should. Create wallet is the only one that has diminished value, but even that could if the incoming pubkey is used for verification

**Pull request merged**